[Accelerate] [vForce] New vForce Overlay

FlexMonkey opened the pull request:
This PR contains a suite of overlays to the vForce transcendental and trigonometric functions on vectors of any length.
The overlays simplify the API to the existing functions and accept collections that implement `_ContiguousCollection`.

This PR contains a Swift-friendly version of each function in vForce, removing the requirement for the developer to separately supply the element count of the source and destination vectors.

cc: @moiseev @airspeedswift @stephentyrone

FlexMonkey committed:
[Accelerate] [vForce] New vForce Overlay

This PR contains a suite of overlays to the vForce transcendental and trigonometric functions on vectors of any length.
The overlays simplify the API to the existing functions and accept collections that implement `_ContiguousCollection`.

FlexMonkey committed:
Change availability version numbers.

FlexMonkey committed:
Change copyright date.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func unbiasedExponent<U, V>(_ vector: U,
```
This should just be `exponent`, and the comment should explain that the value is `floor(log2(vector))`.

FlexMonkey replied:
fixed.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func power<T, U, V>(bases: T,
```
I think we'd prefer to call this `pow`, but the spelling should match whatever the (about to start) review of basic scalar math functions settles on.

FlexMonkey replied:
Changed to `pow` for now - I can always change again depending on the results of your review.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func sinPi<U, V>(_ vector: U,
```
This could also be spelled something like `sin(πTimes:)`.

FlexMonkey replied:
Also did `cosPi` and `tanPi`. I made the parameter name `piTimes` to avoid special characters.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter cosresult: Output cosine values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func sinCos<T, U, V>(_ vector: T,
```
I think this should either use the term of art `sincos` or be a little more explicit: `sinAndCos` or similar.

FlexMonkey replied:
changed to `sincos`.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func arcsin<U, V>(_ vector: U,
```
This should match whatever spelling the math functions proposal settles on (currently `asin`).

FlexMonkey replied:
Changed (also, `acos` and `tan`). Obviously I can change if the proposal settles on something else.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+//
+//===----------------------------------------------------------------------===//
+
+import Accelerate.vecLib
```
Fixed.

FlexMonkey replied:
Fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+
+import Accelerate.vecLib
+
+// Top level enum for namespaced vDSP and vForce based functions
```
Fixed. I've also moved this to its own file - there's no reason (IMHO) why the vForce operations (or any other) should be the special case that houses this top-level enum.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func ceiling<U, V>(_ vector: U,
```
Fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func copySign<T, U, V>(magnitudes: T,
```
Fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func modulus<T, U, V>(numerators: T,
```
Fixed

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func remainder<T, U, V>(numerators: T,
```
Fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func integerTruncation<U, V>(_ vector: U,
```
fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func exponential<U, V>(_ vector: U,
```
fixed.

FlexMonkey commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,1781 @@
+    /// - Parameter result: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func logarithm<U, V>(_ vector: U,
```
fixed.

FlexMonkey committed:
Code Review Fixes,

Mainly renames of functions (e.g. `ceiling` to `ceil`) and parameters (e.g. `numerators` to `dividends`).

FlexMonkey committed:
fix version numbers.

FlexMonkey committed:
Move vForce operations to a vForce namespace.

FlexMonkey committed:
* Create individual `exp`, `expm1`, `exp2` functions.
* Create individual `log2`and `log10` functions.
* Rename `exponent` to `logb`

FlexMonkey committed:
Merge branch 'master' into accelerate-vForce

FlexMonkey committed:
Refactor tests.

FlexMonkey committed:
Add versions of operations that return an array using `unsafeUninitializedCapacity`.

xwu commented:
Based on our [discussion](https://forums.swift.org/t/se-0246-generic-math-s-functions/21479/35) on the basic scalar math functions with `logGamma`, sounds like `sinPi` would be the "less wild" option with precedent in IEEE 754.  Similarly, given the direction that it's taken, `copySign` would be more consistent than `copysign`, and I think `sqrt` is still preferred  over `squareRoot`. By that reasoning, `rSqrt` would be the reciprocal.

Wdyt, @stephentyrone?

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,3435 @@
+            
+            precondition(vector.count == result.count)
+            
+            var n = Int32(vector.count)
```
Can you add a `// TODO:` here to make this work for `count > Int32.max` at some point in the future? It doesn't need to be done now, but it's an obvious enhancement.

FlexMonkey replied:
That would require a change to the Accelerate.vForce API itself where `n` is an `Int32` (for example `[vvceilf](https://developer.apple.com/documentation/accelerate/1470395-vvceilf)`)

stephentyrone replied:
We can handle this for users at the Swift wrapper layer if we want to, by simply calling `vvceilf` in a loop, processing 0x40000000 elements at a go. It's not a huge deal in the short term, because you can't actually allocate a contiguous block of Int32.max floats on any iOS platform (you *can* do it on macOS, however), which is why I'm perfectly content to wait to do this in the future.

FlexMonkey replied:
Todo added.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,3435 @@
+    /// - Returns: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func reciprocalSquareRoot<U>(_ vector: U)  -> [Float]
```
I'm torn between `rsqrt` and `reciprocalSquareRoot` here. I think `rsqrt` may be OK, but if you want to stick with the current verbose spelling, that's also OK with me.

FlexMonkey replied:
If we're changing `squareRoot` to `sqrt`, I think we should use `rsqrt` for consistency.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vForce_Operations.swift:
```diff
@@ -0,0 +1,3435 @@
+    /// - Returns: Output values.
+    @inline(__always)
+    @available(iOS 9999, macOS 9999, tvOS 9999, watchOS 9999, *)
+    public static func squareRoot<U>(_ vector: U)  -> [Float]
```
Make this `sqrt`.

FlexMonkey replied:
Done.

stephentyrone commented:
Keep `copysign`. Switch to `sinPi`, `cosPi`, `sqrt`. I don't love `rSqrt`--I'll leave that one up to @FlexMonkey, but please choose either `reciprocalSquareRoot` or `rsqrt`.

Commit:
Merge branch 'master' into accelerate-vForce

Commit:
Code review changes:

* `squareRoot` -> `sqrt`
* `reciprocalSquareRoot` -> `rsqrt`
* `xxx(piTimes:...)` -> `xxxPi(:_....)`

FlexMonkey commented:
> Keep `copysign`. Switch to `sinPi`, `cosPi`, `sqrt`. I don't love `rSqrt`--I'll leave that one up to @FlexMonkey, but please choose either `reciprocalSquareRoot` or `rsqrt`.

Changed all the `xxx(piTimes:...)` to `xxxPi(_:...)`. 

Cheers.

Commit:
Added todo task.

stephentyrone commented:
I landed this (with some minor changes) as https://github.com/apple/swift/pull/24152. This PR can be closed out.
